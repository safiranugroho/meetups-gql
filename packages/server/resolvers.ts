import { EventResponse } from './data-source';

type Event = {
  name: string;
  date: string;
  time: string;
  venue: string;
  link: string;
};

export default {
  Query: {
    events: async (_, {}, { dataSources: { events } }): Promise<Event[]> => await events.getEvents(),
  },
  Event: {
    date: (event: EventResponse): string => event.local_date,
    time: (event: EventResponse): string => event.local_time,
    venue: (event: EventResponse): string => event.venue.name,
  },
};
