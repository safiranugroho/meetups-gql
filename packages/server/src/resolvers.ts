import moment from 'moment';
import { get } from 'lodash';

import EventsDataSource, { EventResponse } from './data-source';
import GroupsDataSource, { GroupResponse } from './groups-data-source';

export default {
  Query: {
    events: async (
      _: {},
      { input }: { input: { category: number; daysInAdvance: number } },
      {
        dataSources: { events },
      }: { dataSources: { events: EventsDataSource } },
    ) => {
      const category = get(input, 'category');
      const daysInAdvance = get(input, 'daysInAdvance');

      return await events.getEvents(category, daysInAdvance);
    },

    groups: async (
      _: {},
      { input }: { input: { category: number; country: string } },
      {
        dataSources: { groups },
      }: { dataSources: { groups: GroupsDataSource } },
    ) => {
      const category = get(input, 'category');
      const country = get(input, 'country');

      return await groups.getGroups(category, country);
    },
  },
  Event: {
    day: (event: EventResponse): string =>
      moment(get(event, 'local_date')).format('dddd'),
    date: (event: EventResponse): string => get(event, 'local_date'),
    time: (event: EventResponse): string => get(event, 'local_time'),
    epoch: (event: EventResponse): string => get(event, 'time').toString(),
    venue: (event: EventResponse): string => get(event, 'venue.name'),
    group: (event: EventResponse): string => get(event, 'group.name'),
  },
  Group: {
    url: (group: GroupResponse): string => get(group, 'urlname'),
    category: (group: GroupResponse): string => get(group, 'category.name'),
    nextEvent: (group: GroupResponse): EventResponse =>
      get(group, 'next_event'),
  },
};
