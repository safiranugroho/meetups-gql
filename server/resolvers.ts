export default {
  Query: {
    books: (): any[] => [],
  },
};
