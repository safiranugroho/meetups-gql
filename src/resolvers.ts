import { get } from 'lodash';

import EventsDataSource, { EventResponse } from './events-data-source';
import GroupsDataSource, { GroupResponse } from './groups-data-source';

type EventsInput = {
  category?: number;
  daysInAdvance?: number;
};

type GroupsInput = {
  category?: number;
  daysUntilNextEvent?: number;
};

export default {
  Query: {
    events: async (
      _: {},
      { input }: { input?: EventsInput },
      { dataSources }: { dataSources: { events: EventsDataSource } },
    ) => {
      const category = get(input, 'category');
      const daysInAdvance = get(input, 'daysInAdvance');

      return await dataSources.events.getEvents(category, daysInAdvance);
    },

    groups: async (
      _: {},
      { input }: { input?: GroupsInput },
      { dataSources }: { dataSources: { groups: GroupsDataSource } },
    ) => {
      const category = get(input, 'category');
      const daysUntilNextEvent = get(input, 'daysUntilNextEvent');

      return await dataSources.groups.getGroups(category, daysUntilNextEvent);
    },
  },

  Event: {
    date: (event: EventResponse): string => get(event, 'local_date'),
    time: (event: EventResponse): string => get(event, 'local_time'),
    venue: (event: EventResponse): string => get(event, 'venue.name'),
    group: (event: EventResponse): string => get(event, 'group.name'),
  },

  Group: {
    url: (group: GroupResponse): string => get(group, 'urlname'),
    category: (group: GroupResponse): string => get(group, 'category.name'),
    nextEvent: async (
      group: GroupResponse,
      _: {},
      { dataSources }: { dataSources: { groups: GroupsDataSource } },
    ): Promise<EventResponse> =>
      await dataSources.groups.getGroupEvent(
        group.urlname,
        group.next_event.id,
      ),
  },
};
