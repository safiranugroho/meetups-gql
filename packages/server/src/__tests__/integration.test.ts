import nock from 'nock';
import { ApolloServer, gql } from 'apollo-server';
import { createTestClient } from 'apollo-server-testing';

import typeDefs from '../type-defs';
import resolvers from '../resolvers';
import EventsDataSource from '../data-source';

// TODO: Match date string to regex instead of mocking moment
jest.mock('moment', () => ({
  __esModule: true,
  default: () => ({
    add: () => ({
      format: () => '2019-11-10T07:00:00',
    }),
  }),
}));

describe('server', () => {
  const meetupAPI = 'https://api.meetup.com';

  it('should query events according to schema', async () => {
    const GET_EVENTS = gql`
      query($input: EventsInput!) {
        events(input: $input) {
          name
          date
          time
          venue
          link
        }
      }
    `;

    const fakeEventsResponse = {
      events: [
        {
          name: 'fake-event',
          local_date: '2019-11-28',
          local_time: '18:00',
          link: 'http://fake.link',
          venue: {
            name: 'fake-venue',
            city: 'Melbourne',
          },
        },
      ],
    };

    nock(meetupAPI)
      .get('/find/upcoming_events')
      .query({
        topic_category: 292,
        end_date_range: '2019-11-10T07:00:00',
      })
      .reply(200, fakeEventsResponse);

    const server = new ApolloServer({
      typeDefs,
      resolvers,
      dataSources: () => ({
        events: new EventsDataSource(),
      }),
    });

    const { query } = createTestClient(server);
    const response = await query({
      query: GET_EVENTS,
      variables: { input: { category: '292', daysInAdvance: 7 } },
    });

    expect(response).toMatchSnapshot();
  });
});
