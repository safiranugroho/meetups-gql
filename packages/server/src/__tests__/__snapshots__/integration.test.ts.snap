// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server should query events according to schema 1`] = `
Object {
  "data": Object {
    "events": Array [
      Object {
        "date": "2020-04-02",
        "day": "Thursday",
        "group": "Melbourne Silicon Beach",
        "id": "269253587",
        "link": "https://www.meetup.com/Melbourne-Silicon-Beach/events/269253587/",
        "name": "Silicon Beach Pitch Night - April 2020 (Online Event!)",
        "time": "17:30",
        "timeInMilliseconds": "1585809000000",
        "venue": "Online event",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`server should query groups according to schema 1`] = `
Object {
  "data": Object {
    "groups": Array [
      Object {
        "category": "Tech",
        "city": "Melbourne",
        "name": "Melbourne Silicon Beach",
        "nextEvent": Object {
          "date": "2020-04-02",
          "day": "Thursday",
          "group": "Melbourne Silicon Beach",
          "id": "269253587",
          "link": "https://www.meetup.com/Melbourne-Silicon-Beach/events/269253587/",
          "name": "Silicon Beach Pitch Night - April 2020 (Online Event!)",
          "time": "17:30",
          "timeInMilliseconds": "1585809000000",
          "venue": "Online event",
        },
        "url": "Melbourne-Silicon-Beach",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
